### PR TITLE
Sensitivity part II

### LINKED ISSUE #2244: Improve performance of LP sensitivity summary
As raised on Discourse: https://discourse.julialang.org/t/performance-of-lp-rhs-perturbation-range/40551

The function for querying LP sensitivity are pretty slow, mainly because they compute the constraint matrix every call:
https://github.com/JuliaOpt/JuMP.jl/blob/7df1796d305aed88ec4f66ea52b8437fa4babe58/src/lp_sensitivity.jl#L190-L192

There is probably plenty of other low-hanging fruit we could improve.

Edit: in fact, it seems the best thing we could do is compute and return a dictionary with the perturbation range for every constraint.

### PR DESCRIPTION
**Pull request by odow**
Closes #2244

This PR is a rewrite of the sensitivity summary to avoid a lot of redundant computations and allocations.

I've tried to comment my understanding of the code, and compared against the solutions reported by Gurobi and Excel.

Points of contention are:
- What is the sensitivity of a RHS of an `Interval` constraint? I have left unimplemented.
- What is the objective sensitivity of a nonbasic variable with fixed bounds? Gurobi seems to report `+/-0.0`, where as Excel reports `+/-Inf`. The current behavior is `+/-Inf`.

I'm not sure I fully understand all of the edge cases, so it will be interesting to see what the coverage is.

cc @geoffleyland: Want to stress-test this?

### CODE REVIEW

**Comment by joaquimg** [comment]
matrix building is really similar to what is done in:
https://github.com/JuliaOpt/MatrixOptInterface.jl

Matrix building (from MOI) will also be used in the differentiation GSOC.
More general version of that is replicated in solvers like SCS and ProxSDP.

I think the matrix builder should live in MOIU. Starting with LP, for this use, then do QP (for GSOC) and Conic (also fo GSOC, but potentially uses in solvers).

This could simplify some first version solver wrappers.
And could be used by students writing their first solvers.

**Comment by odow** [comment]
:+1: to adding matrix-type support to MOI as a submodule.

File: src/lp_sensitivity2.jl [hunk 1]
```diff
@@ -181,15 +173,14 @@ function lp_sensitivity(model::Model; atol::Float64 = 1e-6)
...
-                # - and nonbasic at the upper bound (switch if lower)
-                # If an even number of these switches have occured, then the
-                # ratio forms an upper bound for δ. Otherwise, it forms a lower
-                # bound.
+                # - and nonbasic at the lower bound (switch if upper)
+                # If an odd number of these switches is true, then the ratio
+                # forms an upper bound for δ. Otherwise, it forms a lower bound.
                 status = j <= n ? variable_status[j] : affine_status[j - n]
-                if iseven(
+                if isodd(
```
Comment (blegat) [inline src/lp_sensitivity2.jl:197]: Looks like `xor(is_min, d[j][e_i] > atol, status == MOI.NONBASIC_AT_LOWER)`

Comment (odow) [reply src/lp_sensitivity2.jl]: I always forget there is a n-ary version.

**Comment by joaquimg** [comment]
I have seen that Xpress allows you to query rhs and obj sensitivities.
Also, Gurobi and Xpress allow you to pass a vector and they will multiply it by the inverse basis matrix.

**Comment by odow** [comment]
Yeah some solvers do allow you to query the sensitivity. But I'm not sure how that plays with bridges, etc.

**Comment by joaquimg** [comment]
Indeed, that is a bit of a pain

**Comment by odow** [comment]
Anyone interested in reviewing this? If not, I'll merge in a few days. It's a significant improvement over what we have currently, which is unusable for anything but toy models.

File: src/lp_sensitivity2.jl [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+            # (δ⁻, δ⁺) = (-πᵢ, ∞) because increasing the objective coefficient
+            # will only keep it at the bound.
+            report.objective[var] = (-π[i], Inf)
+        elseif basis.variables[i] == MOI.NONBASIC_AT_UPPER && !is_min
+            @assert π[i] > -atol
+            # We are maximizing and variable `i` is nonbasic at upper bound.
+            # (δ⁻, δ⁺) = (-πᵢ, ∞) because increasing the objective coefficient
+            # will only keep it at the bound.
+            report.objective[var] = (-π[i], Inf)
+        elseif (
```
Comment (geoffleyland) [inline src/lp_sensitivity2.jl:146]: I don't think you need the parentheses around this condition?

File: src/lp_sensitivity2.jl [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+bounds on `t` such that `x_new` maintains bound feasibility. That is, compute
+bounds on t such that:
+
+    l_B[j] <= x_B[j] + t * d_B[j] <= u_B[j].
+"""
+function _compute_rhs_range(d_B, x_B, l_B, u_B, atol)
+    t_lo, t_hi = -Inf, Inf
+    for j = 1:length(l_B)
+        if d_B[j] > atol
+            t_hi = min(t_hi, (u_B[j] - x_B[j]) / d_B[j])
```
Comment (geoffleyland) [inline src/lp_sensitivity2.jl:255]: In other places in this function, you have `t_lo, t_hi`, but here you do `t_hi` first.  Also, the expressions in both branches are the same, so could compute them beforehand?

File: src/lp_sensitivity2.jl (hunk 2 above)
Comment (odow) [reply src/lp_sensitivity2.jl]: I ran the formatter instead.

File: src/lp_sensitivity2.jl (hunk 3 above)
Comment (odow) [reply src/lp_sensitivity2.jl]: I re-ordered the lines. But I didn't find computing the `t` any more readable. It just added another layer of in-direction.

File: src/lp_sensitivity2.jl [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+"""
+    lp_sensitivity(model::Model; atol::Float64 = 1e-8)::SensitivityReport
+
+Given a linear program `model` containing a current optimal basis, return a
+[`SensitivityReport`](@ref) object, which maps:
+
+ - every constraint reference to a range over which the right-hand side of the
+    corresponding constraint can vary such that the basis remains optimal.
+    ```julia
+    model = lp_sensitivity(model)
```
Comment (joehuchette) [inline src/lp_sensitivity2.jl:31]: Overwriting ``model`` seems incorrect.

File: src/lp_sensitivity2.jl [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+    dRHS_lo, dRHS_hi = model[c]
```
Comment (joehuchette) [inline src/lp_sensitivity2.jl:32]: What is ``c`` here? A ConstraintRef? (Apologies if I'm missing something in how the docs will be rendered)

File: src/lp_sensitivity2.jl [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+            "program (or it contains interval constraints).",
+        )
+    elseif !has_values(model)
+        error("Unable to compute LP sensitivity: no primal solution available.")
+    elseif !has_duals(model)
+        error("Unable to compute LP sensitivity: no dual solution available.")
+    end
+
+    std_form = _standard_form_matrix(model)
+    basis = _standard_form_basis(model, std_form)
```
Comment (joehuchette) [inline src/lp_sensitivity2.jl:59]: Where will this function crash in the case that a basis is not available?

File: src/lp_sensitivity2.jl (hunk 4 above)
Comment (odow) [reply src/lp_sensitivity2.jl:31]: Ooo. I don't know how I missed this one.

File: src/lp_sensitivity2.jl (hunk 6 above)
Comment (odow) [reply src/lp_sensitivity2.jl:59]: It throws an error when querying `ConstraintBasisStatus`.

**Comment by odow** [comment]
Bump. Anyone still want to review this/objections to merging?

**Review by joaquimg — APPROVED** [review]
